vpr/src/route/build_switchblocks.cpp

```diff
@@ -422,6 +432,73 @@ static bool is_core(const DeviceGrid& grid, int x, int y) {
     return is_core;
 }
 
+static bool match_sb_xy(const DeviceGrid& grid, int x, int y, int sb_x, int sb_y, const e_sb_loc_spec& sb_reg_x, const e_sb_loc_spec& sb_reg_y){
+    
+    //if one of sb_x and sb_y is defined, we either know the exact location (x,y) or the exact x location (will apply it to all columns) 
+    //or the exact y location (will apply it to all rows)   
+    if(sb_x != -1 || sb_y != -1){
+        if(x == sb_x && y == sb_y){
+            return true;
+        } 
+
+        if(x == sb_x && sb_y == -1){
+            return true;
+        }
+        
+        if(sb_x == -1 && y == sb_y){
+            return true;
+        }
+    }
+
+    //if both sb_x and sb_y is not defined, we have a region that we should apply this SB pattern to, we just need to check
+    //whether the location passed into this function falls within this region or not
+    if(sb_x == -1 && sb_y == -1){
+        //calculate the appropiate region based on the repeatx/repeaty and current location.
+        //This is to determine whether the given location is part of the current SB specified region with regular expression or not
+        //After region calculation, the current SB will apply to this location if:
+        // 1) the given (x,y) location falls into the calculated region 
+        // *AND*
+        // 2) incrx/incry are respected within the region, this means all locations within the calculated region is
+        //    is not necessary crossponds to the current SB. If incrx/incry is equal to 1, then all locations within the 
+        //    calculated region is valid. 
+        
+        //calculate the region
+        int x_reg_step = (sb_reg_x.repeat != 0) ? (x - sb_reg_x.start) / sb_reg_x.repeat : sb_reg_x.start;
+        int y_reg_step = (sb_reg_y.repeat != 0) ? (y - sb_reg_y.start) / sb_reg_y.repeat : sb_reg_y.start;
+
+        //step must be non-negative
+        x_reg_step = (x_reg_step < 0) ? 0 : x_reg_step;
+        y_reg_step = (y_reg_step < 0) ? 0 : y_reg_step;
+
+        int reg_startx = sb_reg_x.start + (x_reg_step * sb_reg_x.repeat);
+        int reg_endx = sb_reg_x.end + (x_reg_step * sb_reg_x.repeat);
+        reg_endx = std::min(reg_endx, int(grid.width() - 1));
+
+        int reg_starty = sb_reg_y.start + (y_reg_step * sb_reg_y.repeat);
+        int reg_endy = sb_reg_y.end + (y_reg_step * sb_reg_y.repeat);
+        reg_endy = std::min(reg_endy, int(grid.height() - 1));
+  
+        //check x coordinate
+        if (x >= reg_startx && x <= reg_endx){ //should fall into the region
```

Comment:
these nested if statements are a little hard to read
I think the whole check can be implemented in a single if statement